File: app/src/main/java/com/dev/lokeshkalal/creditCard/identifier/service/CardIdentifierServiceImpl.java
package com.dev.lokeshkalal.creditCard.identifier.service;

import com.dev.lokeshkalal.creditCard.Constants;
import com.dev.lokeshkalal.creditCard.CreditCardType;
import com.dev.lokeshkalal.creditCard.identifier.*;

import java.util.ArrayList;

public class CardIdentifierServiceImpl implements CardIdentifierService {

    ArrayList<Identifier> identifiers;

    public CardIdentifierServiceImpl() {
        identifiers = new ArrayList<>();
        identifiers.add(new AmexCardIdentifier());
        identifiers.add(new DinerCardIdentifier());
        identifiers.add(new MasterCardIdentifier());
    }

    @Override
    public CardIdentifierResult identifyCard(String cardnumber) {
        for (Identifier identifier : identifiers) {
            CardIdentifierResult cardIdentifierResult = identifier.validateCard(cardnumber);
            if (cardIdentifierResult.isValidCard()) {
                return cardIdentifierResult;
            }
        }

        return new CardIdentifierResult(CreditCardType.INVALID, Constants.MAX_DEFAULT_CARD_NUMBER, Constants.MAX_DEFAULT_CARD_NUMBER);
    }
}

File: app/src/main/java/com/dev/lokeshkalal/creditCard/identifier/service/CardIdentifierServiceImpl.kt
package com.dev.lokeshkalal.creditCard.identifier.service

import com.dev.lokeshkalal.creditCard.identifier.*

import javax.inject.Inject
import java.util.ArrayList

class CardIdentifierServiceImpl @Inject
constructor() : CardIdentifierService {

    internal var identifiers: ArrayList<Identifier>

    init {
        identifiers = ArrayList()
        identifiers.add(AmexCardIdentifier())
        identifiers.add(DinerCardIdentifier())
        identifiers.add(MasterCardIdentifier())
        identifiers.add(VisaCardIdentifier())
        identifiers.add(JCBCardIdentifier())
        identifiers.add(MaestroCardIdentifier())
        identifiers.add(DiscoverCardIdentifier())
    }

    override fun identifyCard(cardnumber: String): CardIdentifierResult {
        for (identifier in identifiers) {
            val cardIdentifierResult = identifier.validateCard(cardnumber)
            if (cardIdentifierResult.isValidCard) {
                return cardIdentifierResult
            }
        }

        return CardIdentifierResult.InvalidCard()
    }
}
